Misc improvements

c1728p9 opened the pull request:
-Use AHB IDR to determine auto increment size
-Add generic Cortex-M target
-Prevent mon commands from leaving target running
-Add additional registers to GDB server - #114

c1728p9 commented:
```
------ Speed Test Performance ------
Target    Write Speed B/s     Read Speed B/s

kl46z     46217.2001229       46282.4894916
k22f      52787.7558943       52178.3422058
lpc1768   22676.8176566       21977.1974402
k64f      37594.1485925       37385.0537454
lpc800    13429.5052573       13473.6886064
lpc11u24  21729.4513259       20844.7770381



------ TEST RESULTS ------
Target         Test           Result

kl46z          Basic Test     Pass
kl46z          Speed Test     Pass
kl46z          Flash Test     Pass
k22f           Basic Test     Pass
k22f           Speed Test     Pass
k22f           Flash Test     Pass
lpc1768        Basic Test     Pass
lpc1768        Speed Test     Pass
lpc1768        Flash Test     Pass
k64f           Basic Test     Pass
k64f           Speed Test     Pass
k64f           Flash Test     Pass
lpc800         Basic Test     Pass
lpc800         Speed Test     Pass
lpc800         Flash Test     Pass
lpc11u24       Basic Test     Pass
lpc11u24       Speed Test     Pass
lpc11u24       Flash Test     Pass

All tests passed
```

c1728p9 commented on pyOCD/target/cortex_m.py:
```diff
@@ -94,6 +95,14 @@
+                            #                   and arm_cortexm3_processor_trm_100165_0201_00_en.pdf
+    0x44770001 : 0x400,     # Used on m1 - Documented in DDI0413D_cortexm1_r1p0_trm.pdf
+    0x4770031 : 0x400,      # Used on m0+? at least on KL25Z, KL46, LPC812
+    0x4770021 : 0x400,      # Used on m0? used on nrf51, lpc11u24
```
Do you know where these are documented?  For the M1, M3 and M4 the AHB was documented in the respective cortex trm, but for the m0 and m0+ it wasn't.

0xc0170 replied:
I was not able to find them neither (a quick search). We should look at this

0xc0170 commented on pyOCD/flash/flash.py:
```diff
@@ -70,16 +70,22 @@ class Flash(object):
         Download the flash algorithm in RAM
         """
+
+        if self.flash_algo is None:
```
Shall this logging error be only in __init__ method ? It's all over this file, as it can be once there? 

in __init__:

```
if flash_algo is not None:
   #set here all those flash_algo
else:
   error("Target does not have a flash algo")
```

c1728p9 replied:
Rather than in the init, I could put it in flashBinary, flashBlock and getFlashBuilder since those are the primary interface functions for programming.  The only downside would be calling the other functions directly would result in an exception rather than an additional error message.  Let me know what you think.

0xc0170 commented on pyOCD/target/cortex_m.py:
```diff
@@ -267,98 +276,87 @@ class CortexM(Target):
+        RegisterInfo('s29',     32,         'float',        'float'),
+        RegisterInfo('s30',     32,         'float',        'float'),
+        RegisterInfo('s31',     64,         'float',        'float'),
+        ]
```
:+1:

c1728p9 commented:
Updated cortex_m.py to override the core flash functions and thrown an exception if they are called.  Also removed most of the cortex_m logic from flash.py.
